pkg/spiffe/spiffe.go

```diff
@@ -99,6 +101,34 @@ func MustGenSpiffeURI(ns, serviceAccount string) string {
 	return uri
 }
 
+// ExpandWithTrustDomains expands a given spiffe identities, plus a list of truts domain aliases.
+// We ensure the returned list does not contain duplicates; the original input is always retained.
+// For example,
+// ExpandWithTrustDomains({"spiffe://td1/ns/def/sa/def"}, {"td1", "td2"}) returns
+//   {"spiffe://td1/ns/def/sa/def", "spiffe://td2/ns/def/sa/def"}.
+// ExpandWithTrustDomains({"spiffe://td1/ns/def/sa/a", "spiffe://td1/ns/def/sa/b"}, {"td2"}) returns
+//   {"spiffe://td1/ns/def/sa/a", "spiffe://td2/ns/def/sa/a", "spiffe://td1/ns/def/sa/b", "spiffe://td2/ns/def/sa/b"}.
+func ExpandWithTrustDomains(spiffeIdentities, trustDomainAliases []string) map[string]struct{} {
+	out := map[string]struct{}{}
+	for _, id := range spiffeIdentities {
+		out[id] = struct{}{}
+		// Expand with aliases set.
+		m := spiffePattern.FindStringSubmatchIndex(id)
```

Comment:
can we have a function that converts from spiffe to `{TrustDomain, Namespace, ServiceAccount}` that does rely on regex? We will need it for other use cases in the very very near future and I am a bit worried about the regex performance - we call this method a lot

Reply:
AFAIK, we can't assume `spiffe://<td>/ns/<namespace>/sa/<service-account>` format. e.g. the custom identity with annotations.

If the concern is about the regex perf in runtime, I can change it to naive string prefix searching 3rd "/". So perf is more deterministic.

correct way of handling this might be using defined struct? and URI() method when needed. but i'm not sure if that's feasbile given how much spiffe stuff is now everywhere in the code base...

Reply:
Main concern was you doing the work for me to fully parse so I don't have to later in my PR :slightly_smiling_face: . Secondary was possible regex costs at runtime. If we are supporting non spife format, then shouldn't we *not* log an error on L120?

Either way I don't think the fact that we support non spiffe identities means that we should not have a setup like
```go
type SpiffeIdentity struct {
TrustDomain string
Namespace string
ServiceAccount string
}

// Return false if its not spiffe. Or could be an error
func parseSpiffe(s string) (SpiffeIdentity, bool)

func (s SpiffeIdentity) String() string
```

To help isolate the parsing code (easier to test+optimize) and reuse. Not a huge deal though

Reply:
Agree, regex are to be avoided... But not a huge deal for this PR.